cnap/core/keybroker.py

```diff
@@ -0,0 +1,201 @@
+          key (SWK) to encrypt the user key (wrapped_key).
+        * Encrypt the SWK by the public key from client (wrapped_swk).
+    For a key broker client, here is an example flow to get a key from KBS:
+        * Generate 2048 bit RSA key pair (a public key and a private key).
+        * Encode the public key to base64 for transferring (user_data).
+        * Get quote in the TEE with the hash of the public key for measurement (quote).
+        * Request wrapped_key and wrapped_swk from KBS with quote and user_data.
+        * Decrypt the user key by the SWK.
+    """
+    def get_key(self, server_url: str, key_id: str) -> bytes:
+        """Get model key by key ID from the KBS.
+  
+        This method construct the request headers and body to request the wrapped_key and
+        wrapped_swk from KBS, decrypt the user key by SWK and return the key.
+
+        A example requests and response:
+            * request headers:
+                Accept:application/json
+                Content-Type:application/json
+                Attestation-Type:TDX
+            * request body:
+                {
+                    "quote":"",
+                    "user_data":""
+                }
+            * response body:
+                {
+                    "wrapped_key":"",
+                    "wrapped_swk":""
+                }
+
+        Args:
+            server_url (str): The url of KBS.
+            key_id (str): The id of the key.
+
+        Returns:
+            bytes: The bytes of the key.
+    
+        Raises:
+            ValueError: If the server_url or key_id is None.
+            RuntimeError: If get quote or get key failed.
+            NotImplementedError: If the subclasses don't implement the method.
+        """
+        if server_url is None:
+            raise ValueError("KBS server url can not be None")
+        if key_id is None:
+            raise ValueError("KBS key id can not be None")
+
+        private_key = rsa.generate_private_key(public_exponent=65537, key_size=3072)
+        pubkey = private_key.public_key()
+        pubkey_der = pubkey.public_bytes(encoding=serialization.Encoding.DER,
+                                             format=serialization.PublicFormat.SubjectPublicKeyInfo)
+
+        LOG.debug("Getting TDX Quote by CCNP")
+        user_data = base64.b64encode(pubkey_der).decode('utf-8')
+        quote = Quote.get_quote(user_data=user_data)
+        if quote is None:
+            raise RuntimeError("Get TDX Quote failed")
+        quote = base64.b64encode(quote.quote).decode('utf-8')
+
+        req_body = {
+            "quote": quote,
+            "user_data": user_data
+        }
+
+        headers = {
+            "Accept": "application/json",
+            "Content-Type": "application/json",
+            "Attestation-Type": "TDX"
+        }
+
+        LOG.debug("Getting key from the KBS")
+        # retry 3 times for the KBS server issue
+        resp = None
+        for _ in range(3):
```

Comment:
Maybe a `RETRY_TIMES=3` will be more clear?

Reply:
Done.